eosio.system/src/delegate_bandwidth.cpp

```diff
@@ -154,8 +154,17 @@ namespace eosiosystem {
             });
       }
       set_resource_limits( res_itr->owner, res_itr->ram_bytes + ram_gift_bytes, res_itr->net_weight.amount, res_itr->cpu_weight.amount );
+
+      SEND_INLINE_ACTION( *this, notifyram, {payer, N(active)}, {payer, receiver, bytes_out, res_itr->ram_bytes} );
    }
 
+    void system_contract::notifyram(account_name payer, account_name receiver, int64_t bytes_out, int64_t ram_balance)
+    {
+       eosio_assert( bytes_out > 0, "must reserve a positive amount" );
+       eosio_assert( ram_balance > 0, "must reserve a positive amount" );
+       require_recipient(payer);
```

Comment:
this would allow a malicious `receiver` to use any amount of the `payer` s RAM , in reaction to the copy of the action it receives (which carries the authorization of `payer` ).  Until the debate on whether to allow this going forward is settled, it is probably best to avoid using `require_recipient` with anything other than the recipients authority. 

This can be done by dispatching 2 inline actions (one from each authorizer) and then in the `system_contract::notifyram` action using `has_auth` to determine which require_recipient to call.

Reply:
If I understood you correctly:

- In v1.2.3 release already deprecated of the ability of a notified contract to bill RAM to authorizers of the original action, if `receiver` do tries to consume `payer`'s RAM, the transaction will fail.
- Your suggested solution is to send `notifyram` inline actions to `payer` and `receiver` respectively, and use `has_auth` inside `notifyram` to decide that party have the authority to receive recipient?

Reply:
@wanderingbort, @chengevo:

Since the `notifyram` action doesn't actually `require_auth`, it could also be possible to just leave off the `payer@active` authorization in the inline action that is sent from `buyram`. Then there is no authorizer that can be charged RAM by the notified parties.

However, this exposes another problem with the current code which is that the notified contracts cannot trust that this `notifyram` action is authentic (meaning only produced by the system contract in response to a `buyram` action). Someone could spoof it which may break the invariants in the notified contracts.

One way to fix that is to pass in the authorization of `eosio@active` (or alternatively to play it even more safe while still providing spoofing resistance pass in the authorization of, for example, `eosio.ram@active`) when sending the inline action, and to use `require_auth( _self );` (or `require_auth( N(eosio.ram) );`) in `notifyram` to ensure it has the proper authorization.

However, this fix then makes the RAM problem much worse, because then either of the notified recipients can charge RAM to `eosio` (or `eosio.ram` which is only slightly better)! This means such a change to the system contract should be delayed until the code has been released for the protocol upgrade that objectively enforces the notify RAM changes that can as of v1.2.3 only be subjectively enforced. And even then the system contract code would need to utilize the privileged `is_feature_active` intrinsic to assert in `notifyram` that the objective check is in place.

In my opinion, I don't contracts should even bother trying to track the RAM purchases of accounts. Perhaps some off-chain clients would be interested in this, but it is already in theory possible to do this of course.

I can see an argument that it should be made easier for off-chain clients and even add some cryptographic assurances on top by instead dispatching an inline action (to simply act as an event for off-chain consumers) with the results of the RAM purchase. The action would still `require_auth( _self );` (or `require_auth( N(eosio.ram) );`) to enforce authenticity of the event but it wouldn't call `require_recipient` at all.